### PR TITLE
Accept SSH environment variable passing

### PR DESCRIPTION
**Pull request by cjwatson**
## Contributor Checklist:

* [x] There is an associated ticket in Trac. https://twistedmatrix.com/trac/ticket/9315
* [x] The changes pass minimal style checks
* [x] I have created a newsfragment in src/twisted/conch/newsfragments/
* [x] I have updated the automated tests.

### CODE REVIEW

File: src/twisted/conch/interfaces.py
Comment (twm) [inline src/twisted/conch/interfaces.py]: Unfortunately this isn't backwards compatible. [Interfaces are forever.](https://orbifold.xyz/interfaces-are-forever.html) You will have to create a new interface to host this method, e.g. `ISetEnv`, and check if the session object implements it in `request_env`.

File: src/twisted/conch/interfaces.py
Comment (twm) [inline src/twisted/conch/interfaces.py]: It would be useful to mention what the OpenSSH-compatible behavior is here, since any alternative implementation will probably want to implement it.

File: src/twisted/conch/ssh/session.py
Comment (twm) [inline src/twisted/conch/ssh/session.py:132]: Please add `@param` and `@returns` as well, per the [coding standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html).

File: src/twisted/conch/test/test_session.py
Comment (twm) [inline src/twisted/conch/test/test_session.py]: This test doesn't seem to cover the `ValueError` code path that silently ignores a variable. We need that coverage.

**Review by twm — CHANGES_REQUESTED** [review]
Hi Colin,

This is a great feature, and it's great to see it happening! I noted a few things in the review comments below. You'll need to fix the incompatible interface change and add the missing test coverage before we can land.  I'll add that I'm not too familiar with Conch myself, so I'm mostly looking at this from a general Twisted background and might miss Conch-specific stuff.

The ticket mentions implementing OpenSSH's environment filtering behavior. It seems like Conch's default server should do this, though if you prefer to file a follow-up ticket for that it's fine.

Thanks for contributing to Twisted!

File: src/twisted/conch/interfaces.py
Comment (cjwatson) [reply src/twisted/conch/interfaces.py]: So, I understand that argument and I'm willing to do this, but I thought I'd first explain why I did it this way and see whether it changes your mind.

Generally speaking I try to fit into the style of existing code before almost any other consideration.  In this case, all of the `request_*` methods in `SSHSession` wrap their calls to the session object in very general exception handlers.  (I couldn't quite bring myself to imitate the bare `except`s that they use, but I imitated the rest of it.)  As such, it is effectively already true that `request_env` checks whether the session object implements the necessary method: if it doesn't, it results in an error being logged, but that's as far as it goes.  Of the current `ISession` implementers in Twisted itself, `docs/conch/examples/sshsimpleserver.py:ExampleSession`, `twisted.conch.manhole_ssh.TerminalSession`, and `twisted.conch.test.test_conch.RekeyAvatar` already fail to implement all of the methods in `ISession`, so for better or worse the current code can't be being strict about it.

Now, I can see that there might be an argument that `ISession` is a public interface and so external code is entitled to depend on it in other ways; but the above is why I took this approach in the first pass.

File: src/twisted/conch/interfaces.py
Comment (cjwatson) [reply src/twisted/conch/interfaces.py]: Done.

File: src/twisted/conch/ssh/session.py
Comment (cjwatson) [reply src/twisted/conch/ssh/session.py:132]: Done.

File: src/twisted/conch/test/test_session.py
Comment (cjwatson) [reply src/twisted/conch/test/test_session.py]: Done.

**Comment by cjwatson** [comment]
Regarding the behaviour of Conch's default server: OpenSSH's default behaviour is to reject all environment variables, and the only default server that I know of that Conch ships is `twisted.conch.tap` which only has very minimal configuration.  My understanding is that the typical way to use Conch as a server is to use it as a library of components from which you can assemble a server with the custom behaviour you want (that's certainly how we use it in Launchpad).  But I could be wrong: perhaps @markrwilliams can elaborate on their requirements?

Note that the cited behaviour of OpenSSH in accepting `LANG` and `LC_*` is *not* the default behaviour upstream, but rather a [Debian-specific customisation](https://manpages.debian.org/stable/openssh-server/sshd_config.5.en.html).

File: src/twisted/conch/interfaces.py
Comment (twm) [reply src/twisted/conch/interfaces.py]: > Now, I can see that there might be an argument that `ISession` is a public interface and so external code is entitled to depend on it in other ways; but the above is why I took this approach in the first pass.

Yeah, that's pretty much it. I don't view it so much as an argument as Twisted policy. But maybe this should be an exception if none of the real-world implementations actually _implement_ the interface. This seems like a question for the mailing list &mdash; again, I'm not too familiar with Conch.

Comment (cjwatson) [reply src/twisted/conch/interfaces.py]: It's OK, I just wanted to get an idea of how firm the requirement was.  I've moved this to a new `ISessionSetEnv` interface (which necessitated some test rearrangements to be able to effectively test the case where the avatar doesn't implement that interface).

**Comment by glyph** [comment]
FWIW I’d love it if Conch’s `tap` command line were a *bit* more robust in terms of supporting this sort of configuration.

File: src/twisted/conch/ssh/session.py
Comment (glyph) [inline src/twisted/conch/ssh/session.py]: It makes me a bit sad to see a new old-style log message get added like this 😞 .  But I guess we don't have a great "list of migrations in progress" document that reviewers should look at, so it's not surprising it got missed.

To make a new-style log message, make a new `twisted.logger.Logger` instance at module scope, and call `log.<appropriate level method>`.  In this case, probably `log.info`.

Also: unhandled request for *what* env?  Of whom?  All log messages should include relevant information that will make them actionable by administrators, or at least by the implementor of the avatar class.  In this case I'd say something like `_log.info('SSH avatar {avatar} does not handle setting requested environment variables {env}', avatar=self.avatar, env=data)` or similar.

File: src/twisted/conch/ssh/session.py
Comment (glyph) [inline src/twisted/conch/ssh/session.py]: A better way to address the lack of type information than catching `TypeError` (which may be raised by a great many things in Python's runtime or standard library) is to do `ISessionSetEnv(self.avatar, None)`.

File: src/twisted/conch/ssh/session.py
Comment (glyph) [inline src/twisted/conch/ssh/session.py]: This interface is already somewhat over-broad; is it really necessary to make `sessionSetEnv` a public read/write variable here?  If so, we definitely need to document it for callers.  I mean, it'd be nice to have documentation of its type either way, but an underscore prefix means that documentation is only really maintainer-facing rather than user-facing ;)

File: src/twisted/conch/ssh/session.py
Comment (glyph) [inline src/twisted/conch/ssh/session.py:68]: (Thanks for the cleanup.)

File: src/twisted/conch/interfaces.py
Comment (glyph) [inline src/twisted/conch/interfaces.py]: `ValueError` seems like a pretty broad exception to catch with no logging at all.  For example, if your application code is expecting an `int`-able value and your client transmitted `"1.0"`, you'll get a `ValueError` and it'll get swallowed.

In general, this is why using stdlib exceptions is generally a bad idea for specific failure modes.

Could we define a new exception here, `EnvironmentVariableNotPermitted`, so it's clear that it's being raised intentionally?  Alternately, if the only two possible behaviors here are "variable was set, OK" and "variable rejected", perhaps the return value ought to be an enum or something similar, and exceptions shouldn't be allowed in the interface?

File: src/twisted/conch/newsfragments/9315.feature
Comment (glyph) [inline src/twisted/conch/newsfragments/9315.feature]: 👍 Love the description.  Could you qualify `ISessionSetEnv` and perhaps be a little more specific than "using" though?  Like "by your SSH avatar implementing"?

File: src/twisted/conch/test/test_session.py
Comment (glyph) [inline src/twisted/conch/test/test_session.py]: This seems like a *suspiciously* specific value to choose for a stub test :).  Would you mind making it a less "set off security scanners" type value? :)

File: src/twisted/conch/ssh/session.py
Comment (glyph) [inline src/twisted/conch/ssh/session.py]: This seems like a good candidate for the "active nothing" pattern, where it's never `None` but sometimes it's a specialized no-op implementation that refuses all variables?  This pushes some of the conditional logic below into the stub implementation, making it less complex.

File: src/twisted/conch/ssh/session.py
Comment (glyph) [inline src/twisted/conch/ssh/session.py]: As above re: new vs. old style loggers: with a new style logger this would be `_log.failure()`

File: src/twisted/conch/test/test_session.py
Comment (glyph) [inline src/twisted/conch/test/test_session.py:83]: Thanks for the docs!

**Review by glyph — CHANGES_REQUESTED** [review]
Sorry to have a secondary review that piles on more feedback, but there are a couple of public-interface concerns here which I would really like to see addressed before landing.  Thanks again for doing this, this is a hugely useful feature for conch!

When you have addressed feedback, feel free to @ me here, and I'll do my best to re-review promptly to get this integrated.

File: src/twisted/conch/ssh/session.py
Comment (cjwatson) [reply src/twisted/conch/ssh/session.py]: I'd prefer not to do this in this case, because that loses the distinction for logging purposes between "can't set environment variables because doing that isn't implemented" and "can't set environment variables because the implementation said no", which seems to me to be likely to be useful.  (Or, if I were to preserve the extra logging, as far as I can see it would make no difference to the complexity of the conditional logic below it: it still needs to cope with the three cases of OK, explicitly-forbidden, and error.)

**Comment by cjwatson** [comment]
@glyph Thanks for the review.  I think I've addressed all your comments in one way or another; please re-review?

File: src/twisted/conch/ssh/session.py
Comment (ryban) [reply src/twisted/conch/ssh/session.py]: Is there a particular reason to place the log at the module scope? It would be more helpful to me for the logger to be at a class attribute so the exact instance can be known from `log_source` in the event. In my case I want to associate logs from a certain SSH connection to the "owner" of that connection.

**Review by twm — APPROVED** [review]
Indeed, all the review feedback has been addressed. I will go ahead and merge (the test failure is a recurring Azure issue unrelated to these changes). Thank you very much!

File: src/twisted/conch/ssh/session.py
Comment (twm) [reply src/twisted/conch/ssh/session.py]: @ryban I think we can address that as part of [#9773](https://twistedmatrix.com/trac/ticket/9773)

**Pull request merged**
